File: leetcode/python/997_find_town_judge.py
from collections import defaultdict

class Solution:
    def findJudge(self, n: int, trust: list[list[int]]) -> int:
        nontrusters = set([i for i in range(1, n+1)])
        trustCount = defaultdict(int)
        for [truster, trustee] in trust:
            if truster in nontrusters:
                nontrusters.remove(truster)
            trustCount[trustee] += 1
        if len(nontrusters) == 1 and trustCount[(i := nontrusters.pop())] == n - 1:
            return i
        return -1

if __name__ == '__main__':
    s = Solution()
    print(s.findJudge(3, [[1,3],[2,3]]))
    print(s.findJudge(3, [[1,3],[2,3],[3,1]]))
    print(s.findJudge(4, [[1,3],[1,4],[2,3],[2,4],[4,3]]))

File: leetcode/rust/src/997_find_town_judge.rs
impl Solution {
    pub fn find_judge(n: i32, trust: Vec<Vec<i32>>) -> i32 {
        let mut trust_count = vec![0; n as usize];
        for t in trust {
            trust_count[t[0] as usize - 1] -= 1;
            trust_count[t[1] as usize - 1] += 1;
        }
        for i in 0..n as usize {
            if trust_count[i] == n - 1 {
                return i as i32 + 1;
            }
        }
        -1
    }
}